src/operators/groupBy.ts

```diff
@@ -0,0 +1,131 @@
+{
+  return this.lift(new GroupByOperator<T, R, K>(keySelector, durationSelector, elementSelector));
+}
+
+export class GroupByOperator<T, R, K> extends Operator<T, R> {
+  constructor(private keySelector: (value: T) => K, 
+    private durationSelector?: (grouped: GroupSubject<R, K>) => Observable<any>,
+    private elementSelector?: (value: T) => R) {
+    super();
+  }
+  
+  call(observer: Observer<R>): Observer<T> {
+    return new GroupBySubscriber<T, R, K>(observer, this.keySelector, this.durationSelector, this.elementSelector);
+  }
+}
+
+export class GroupBySubscriber<T, R, K> extends Subscriber<T> {
+  public groups = {};
+  
+  constructor(destination: Observer<R>, private keySelector: (value: T) => K,
+    private durationSelector?: (grouped: GroupSubject<R, K>) => Observable<any>, 
+    private elementSelector?: (value: T) => R) {
+    super(destination);
+  }
+
+  _next(x: T) {
+    let key = tryCatch(this.keySelector)(x);
+    if(key === errorObject) {
+      this.error(key.e);
+    } else {
+      const groups = this.groups;
+      const elementSelector = this.elementSelector;
+      const durationSelector = this.durationSelector;
+      
+      let group: GroupSubject<R, K> = groups[key];
+    
+      if (!group) {
+        groups[key] = group = new GroupSubject(key);
+        
+        if (durationSelector) {
+          let duration = tryCatch(durationSelector)(group);
+          if (duration === errorObject) {
+            this.error(duration.e);
+          } else {
+            group.duration = duration;
+            this.add(duration.subscribe(new GroupDurationSubscriber(group, this)));
+          }
+        }
+        
+        this.destination.next(group);
+      }
+      
+      if (elementSelector) {
+        let value = tryCatch(elementSelector)(x)
+        if(value === errorObject) {
+          group.error(value.e);
+        } else {
+          group.next(value);
+        }
+      } else {
+        group.next(x);
+      }
+    }
+  }
+  
+  _error(err: any) {
+    const groups = this.groups;
+    for (let key in groups) {
+      let group = groups[key];
+      if (group) {
+        group.error(err);
+        groups[key] = null;
+      }
+    }
+    this.destination.error(err);
+  }
+  
+  _complete() {
+    const groups = this.groups;
+    for(let key in groups) {
+      let group = groups[key];
+      if(group) {
+        group.complete();
+        groups[key] = null;
+      }
+    }
+    this.destination.complete();
+  }
+}
+
+export class GroupSubject<T, K> extends Subject<T> {
+  duration: Observable<any>;
+  
+  constructor(public key: K) {
+    super();
+  } 
+}
+
+export class GroupDurationSubscriber<T> extends Subscriber<T> {
+  constructor(private group: GroupSubject<T, any>, private parent:GroupBySubscriber<any, T, any>) {
+    super(null);
+  }
+  
+  _next(value: T) {
+    this.group.complete();
+    this.parent.groups[this.group.key] = null;
```

Comment:
Nitpicking here, but this logic seems ripe to put into a function on the GroupBySubscriber, the way we've done it in [merge](https://github.com/ReactiveX/RxJS/blob/master/src/operators/merge.ts#L106). This would also make it easier to subclass.